File: bridge/worker_pool.py
"""Worker Pool Module - CollectiveBrain Multi-Agent System

Specialized worker agents that execute assigned subtasks with deterministic IDs.
Follows the Orchestrator-Worker pattern as defined in the Shared Constitution.
"""

from typing import Dict, Any, Optional
from datetime import datetime
import uuid


class WorkerAgent:
    """Base worker agent that executes assigned subtasks."""
    
    def __init__(self, role: str, agent_id: Optional[str] = None):
        self.role = role
        self.agent_id = agent_id or str(uuid.uuid4())
        self.task_history: list = []
        self.current_task: Optional[Dict] = None
    
    def execute_task(self, task_id: str, instruction: str) -> Dict[str, Any]:
        """
        Execute a subtask assigned by the orchestrator.
        
        Args:
            task_id: Unique task identifier
            instruction: Specific instruction for this subtask
        
        Returns:
            Execution result with structured output
        """
        start_time = datetime.utcnow().isoformat()
        
        self.current_task = {
            "task_id": task_id,
            "instruction": instruction,
            "status": "in_progress",
            "started_at": start_time
        }
        
        # Placeholder: Later route to specific tools/models based on role
        # This is where GitHub Models API calls would be integrated
        result_data = f"Stub result from {self.role} agent for: {instruction}"
        
        result = {
            "task_id": task_id,
            "agent_id": self.agent_id,
            "role": self.role,
            "instruction": instruction,
            "result": result_data,
            "status": "completed",
            "started_at": start_time,
            "completed_at": datetime.utcnow().isoformat(),
            "reflection_token": "[IsRel]"  # Reflection token for self-correction
        }
        
        self.task_history.append(result)
        self.current_task = None
        
        return result
    
    def get_status(self) -> Dict[str, Any]:
        """Get current agent status."""
        return {
            "agent_id": self.agent_id,
            "role": self.role,
            "current_task": self.current_task,
            "tasks_completed": len(self.task_history),
            "is_available": self.current_task is None
        }


class WorkerPool:
    """Manages a pool of specialized worker agents."""
    
    def __init__(self):
        self.workers: Dict[str, WorkerAgent] = {}
        self.worker_roles = ["Research", "Finance", "Analysis", "Implementation"]
        self._initialize_workers()
    
    def _initialize_workers(self):
        """Initialize default worker agents for each role."""
        for role in self.worker_roles:
            worker = WorkerAgent(role=role)
            self.workers[worker.agent_id] = worker
    
    def get_available_worker(self, role: str) -> Optional[WorkerAgent]:
        """Get an available worker by role."""
        for worker in self.workers.values():
            if worker.role == role and worker.current_task is None:
                return worker
        return None
    
    def assign_task(self, role: str, task_id: str, instruction: str) -> Dict[str, Any]:
        """
        Assign a task to an available worker of the specified role.
        
        Args:
            role: Worker role (e.g., 'Research', 'Finance')
            task_id: Unique task identifier
            instruction: Task instruction
        
        Returns:
            Task execution result
        """
        worker = self.get_available_worker(role)
        
        if not worker:
            return {
                "error": f"No available worker for role: {role}",
                "status": "failed"
            }
        
        return worker.execute_task(task_id, instruction)
    
    def get_pool_status(self) -> Dict[str, Any]:
        """Get status of all workers in the pool."""
        return {
            "total_workers": len(self.workers),
            "available_workers": sum(1 for w in self.workers.values() if w.current_task is None),
            "workers": [w.get_status() for w in self.workers.values()]
        }


if __name__ == "__main__":
    # Example usage
    pool = WorkerPool()
    
    print(f"Pool initialized with {len(pool.workers)} workers")
    print(f"\nAvailable roles: {pool.worker_roles}")
    
    # Assign a task
    result = pool.assign_task("Research", "task-123", "Research vector database options")
    print(f"\nTask result: {result['status']}")
    print(f"Result data: {result['result']}")
    print(f"Reflection token: {result['reflection_token']}")

File: services/trinity/packages/brain/src/core/worker-pool.js
/**
 * Worker Pool - OMEGA Core
 * 
 * Specialized worker agents that execute assigned subtasks.
 * Ported from Python bridge/worker_pool.py
 * 
 * Follows the Orchestrator-Worker pattern.
 */

const { v4: uuidv4 } = require('uuid');

/**
 * Base worker agent that executes assigned subtasks
 */
class WorkerAgent {
    constructor(role, agentId = null) {
        this.role = role;
        this.agentId = agentId || uuidv4();
        this.taskHistory = [];
        this.currentTask = null;
    }

    /**
     * Execute a subtask assigned by the orchestrator
     */
    executeTask(taskId, instruction) {
        const startTime = new Date().toISOString();

        this.currentTask = {
            task_id: taskId,
            instruction,
            status: 'in_progress',
            started_at: startTime
        };

        // Placeholder: Later route to specific tools/models based on role
        const resultData = `Stub result from ${this.role} agent for: ${instruction}`;

        const result = {
            task_id: taskId,
            agent_id: this.agentId,
            role: this.role,
            instruction,
            result: resultData,
            status: 'completed',
            started_at: startTime,
            completed_at: new Date().toISOString(),
            reflection_token: '[IsRel]' // Reflection token for self-correction
        };

        this.taskHistory.push(result);
        this.currentTask = null;

        return result;
    }

    /**
     * Get current agent status
     */
    getStatus() {
        return {
            agent_id: this.agentId,
            role: this.role,
            current_task: this.currentTask,
            tasks_completed: this.taskHistory.length,
            is_available: this.currentTask === null
        };
    }
}

/**
 * Manages a pool of specialized worker agents
 */
class WorkerPool {
    constructor() {
        this.workers = {};
        this.workerRoles = ['Research', 'Finance', 'Analysis', 'Implementation'];
        // Back-compat alias
        this.worker_roles = this.workerRoles;
        this._initializeWorkers();
    }

    _initializeWorkers() {
        for (const role of this.workerRoles) {
            const worker = new WorkerAgent(role);
            this.workers[worker.agentId] = worker;
        }
    }

    /**
     * Get an available worker by role
     */
    getAvailableWorker(role) {
        for (const worker of Object.values(this.workers)) {
            if (worker.role === role && worker.currentTask === null) {
                return worker;
            }
        }
        return null;
    }

    /**
     * Assign a task to an available worker of the specified role
     */
    assignTask(role, taskId, instruction) {
        const worker = this.getAvailableWorker(role);

        if (!worker) {
            return {
                error: `No available worker for role: ${role}`,
                status: 'failed'
            };
        }

        return worker.executeTask(taskId, instruction);
    }

    /**
     * Get status of all workers in the pool
     */
    getPoolStatus() {
        return {
            total_workers: Object.keys(this.workers).length,
            available_workers: Object.values(this.workers).filter(w => w.currentTask === null).length,
            workers: Object.values(this.workers).map(w => w.getStatus())
        };
    }
}

module.exports = { WorkerAgent, WorkerPool };
